File: app/src/main/java/com/kimjisub/launchpad/activity/FBStoreActivity.java
package com.kimjisub.launchpad.activity;

import android.annotation.SuppressLint;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;
import com.kimjisub.design.PackViewSimple;
import com.kimjisub.launchpad.BuildConfig;
import com.kimjisub.launchpad.R;
import com.kimjisub.launchpad.adapter.StoreAdapter;
import com.kimjisub.launchpad.adapter.StoreItem;
import com.kimjisub.launchpad.databinding.ActivityStoreBinding;
import com.kimjisub.launchpad.manager.PreferenceManager;
import com.kimjisub.launchpad.manager.Unipack;
import com.kimjisub.launchpad.network.Networks;
import com.kimjisub.launchpad.network.fb.StoreVO;
import com.kimjisub.manager.FileManager;
import com.kimjisub.manager.Log;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.ConcurrentModificationException;

public class FBStoreActivity extends BaseActivity {
	ActivityStoreBinding b;

	Networks.FirebaseManager firebase_store;
	Networks.FirebaseManager firebase_storeCount;

	ArrayList<StoreItem> list;
	StoreAdapter adapter;
	File[] F_UniPackList;

	void initVar(boolean onFirst) {
		if (onFirst) {
			firebase_store = new Networks.FirebaseManager("store");
			firebase_storeCount = new Networks.FirebaseManager("storeCount");
			list = new ArrayList<>();
			adapter = new StoreAdapter(FBStoreActivity.this, list, new StoreAdapter.EventListener() {

				@Override
				public void onViewClick(StoreItem item, PackViewSimple v) {
					togglePlay(item);
				}

				@Override
				public void onViewLongClick(StoreItem item, PackViewSimple v) {

				}

				@Override
				public void onPlayClick(StoreItem item, PackViewSimple v) {
					if (!item.isDownloaded() && !item.isDownloading())
						startDownload(getPackItemByCode(item.getStoreVO().code));
				}
			});
			adapter.registerAdapterDataObserver(new RecyclerView.AdapterDataObserver() {
				@Override
				public void onItemRangeInserted(int positionStart, int itemCount) {
					super.onItemRangeInserted(positionStart, itemCount);
					b.errItem.setVisibility(list.size() == 0 ? View.VISIBLE : View.GONE);
				}

				@Override
				public void onItemRangeRemoved(int positionStart, int itemCount) {
					super.onItemRangeRemoved(positionStart, itemCount);
					b.errItem.setVisibility(list.size() == 0 ? View.VISIBLE : View.GONE);
				}
			});

			DividerItemDecoration divider = new DividerItemDecoration(FBStoreActivity.this, DividerItemDecoration.VERTICAL);
			divider.setDrawable(getResources().getDrawable(R.drawable.border_divider));
			b.recyclerView.addItemDecoration(divider);
			b.recyclerView.setHasFixedSize(false);
			b.recyclerView.setLayoutManager(new LinearLayoutManager(FBStoreActivity.this));
			//b.recyclerView.setItemAnimator(null);
			b.recyclerView.setAdapter(adapter);
		}
		F_UniPackList = getUniPackDirList();
	}

	// =============================================================================================

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		b = setContentViewBind(R.layout.activity_store);
		initVar(true);

		firebase_store.setEventListener(new ChildEventListener() {
			@Override
			public void onChildAdded(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {
				Log.test("onChildAdded: " + s);

				try {
					StoreVO d = dataSnapshot.getValue(StoreVO.class);

					boolean isDownloaded = false;
					for (File dir : F_UniPackList) {
						if (d.code.equals(dir.getName())) {
							isDownloaded = true;
							break;
						}
					}


					list.add(0, new StoreItem(d, isDownloaded));
					adapter.notifyItemInserted(0);


					updatePanelMain();
				} catch (Exception e) {
					e.printStackTrace();
				}
				Log.test("onChildAdded: ");
			}

			@Override
			public void onChildChanged(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {
				Log.test("onChildChanged: " + s);

				try {
					StoreVO d = dataSnapshot.getValue(StoreVO.class);
					StoreItem item = getPackItemByCode(d.code);
					item.setStoreVO(d);
					adapter.notifyItemChanged(list.indexOf(item), "update");

					int selectedIndex = getSelectedIndex();
					if (selectedIndex != -1) {
						String changeCode = item.getStoreVO().code;
						String selectedCode = list.get(selectedIndex).getStoreVO().code;
						if (changeCode.equals(selectedCode))
							updatePanelPack(item);
					}


				} catch (Exception e) {
					e.printStackTrace();
				}
			}

			@Override
			public void onChildRemoved(@NonNull DataSnapshot dataSnapshot) {
			}

			@Override
			public void onChildMoved(@NonNull DataSnapshot dataSnapshot, @Nullable String s) {
			}

			@Override
			public void onCancelled(@NonNull DatabaseError databaseError) {
			}
		});

		firebase_storeCount.setEventListener(new ValueEventListener() {
			@Override
			public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
				Long data = dataSnapshot.getValue(Long.class);
				PreferenceManager.PrevStoreCount.save(FBStoreActivity.this, data);
			}

			@Override
			public void onCancelled(@NonNull DatabaseError databaseError) {

			}
		});
	}

	// ============================================================================================= List Manage

	void togglePlay(StoreItem target) {
		try {
			for (StoreItem item : list) {
				PackViewSimple packViewSimple = item.getPackViewSimple();

				if (target != null && item.getStoreVO().code.equals(target.getStoreVO().code))
					item.setToggle(!item.isToggle());
				else
					item.setToggle(false);

				if (packViewSimple != null)
					packViewSimple.toggle(item.isToggle());
			}

			updatePanel();

		} catch (ConcurrentModificationException e) {
			e.printStackTrace();
		}
	}

	int getSelectedIndex() {
		int index = -1;

		int i = 0;
		for (StoreItem item : list) {
			if (item.isToggle()) {
				index = i;
				break;
			}
			i++;
		}

		return index;
	}

	StoreItem getPackItemByCode(String code) {
		StoreItem ret = null;
		for (StoreItem item : list)
			if (item.getStoreVO().code.equals(code)) {
				ret = item;
				break;
			}
		return ret;
	}

	int getDownloadingCount() {
		int count = 0;
		for (StoreItem item : list) {
			if (item.isDownloading())
				count++;
		}

		return count;
	}

	int getDownloadedCount() {
		int count = 0;
		for (StoreItem item : list) {
			if (item.isDownloaded())
				count++;
		}

		return count;
	}

	@SuppressLint("StaticFieldLeak")
	void startDownload(StoreItem item) {
		StoreVO StoreVO = item.getStoreVO();
		PackViewSimple packViewSimple = item.getPackViewSimple();

		String code = StoreVO.code;
		String title = StoreVO.title;
		String producerName = StoreVO.producerName;
		boolean isAutoPlay = StoreVO.isAutoPlay;
		boolean isLED = StoreVO.isLED;
		int downloadCount = StoreVO.downloadCount;
		String URL = StoreVO.URL;

		File F_UniPackZip = FileManager.makeNextPath(F_UniPackRootExt, code, ".zip");
		File F_UniPack = new File(F_UniPackRootExt, code);

		packViewSimple.animateFlagColor(color(R.color.gray1));
		packViewSimple.setPlayText("0%");

		(new AsyncTask<String, Long, String>() {


			@Override
			protected void onPreExecute() {
				item.setDownloading(true);

				super.onPreExecute();
			}

			@Override
			protected String doInBackground(String[] params) {

				Networks.sendGet("https://us-central1-unipad-e41ab.cloudfunctions.net/increaseDownloadCount/" + code);
				try {

					URL url = new URL(URL);
					HttpURLConnection connection = (HttpURLConnection) url.openConnection();
					connection.setConnectTimeout(5000);
					connection.setReadTimeout(5000);

					int fileSize = connection.getContentLength();
					Log.log(URL);
					Log.log("fileSize : " + fileSize);
					fileSize = fileSize == -1 ? 104857600 : fileSize;

					InputStream input = new BufferedInputStream(url.openStream());
					OutputStream output = new FileOutputStream(F_UniPackZip);

					byte[] data = new byte[1024];

					long total = 0;

					int count;
					int skip = 100;
					while ((count = input.read(data)) != -1) {
						total += count;
						skip--;
						if (skip == 0) {
							publishProgress(0L, total, (long) fileSize);
							skip = 100;
						}
						output.write(data, 0, count);
					}

					output.flush();
					output.close();
					input.close();
					publishProgress(1L);

					try {
						FileManager.unZipFile(F_UniPackZip.getPath(), F_UniPack.getPath());
						Unipack unipack = new Unipack(F_UniPack, true);
						if (unipack.CriticalError) {
							Log.err(unipack.ErrorDetail);
							publishProgress(-1L);
							FileManager.deleteDirectory(F_UniPack);
						} else
							publishProgress(2L);

					} catch (Exception e) {
						publishProgress(-1L);
						FileManager.deleteDirectory(F_UniPack);
						e.printStackTrace();
					}
					FileManager.deleteDirectory(F_UniPackZip);


				} catch (Exception e) {
					publishProgress(-1L);
					e.printStackTrace();
				}

				item.setDownloading(false);

				return null;
			}

			@Override
			protected void onProgressUpdate(Long... progress) {
				if (progress[0] == 0) {//다운중
					packViewSimple.setPlayText((int) ((float) progress[1] / progress[2] * 100) + "%\n" + FileManager.byteToMB(progress[1]) + " / " + FileManager.byteToMB(progress[2]) + "MB");
				} else if (progress[0] == 1) {//분석중
					packViewSimple.setPlayText(lang(R.string.analyzing));
					packViewSimple.animateFlagColor(color(R.color.orange));
				} else if (progress[0] == -1) {//실패
					packViewSimple.setPlayText(lang(R.string.failed));
					packViewSimple.animateFlagColor(color(R.color.red));
				} else if (progress[0] == 2) {//완료
					packViewSimple.setPlayText(lang(R.string.downloaded));
					packViewSimple.animateFlagColor(color(R.color.green));
					item.setDownloaded(true);
					updatePanel();
				}
			}

			@Override
			protected void onPostExecute(String unused) {

			}
		}).executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
	}

	// ============================================================================================= panel

	void updatePanel() {
		Log.test("panel");
		int playIndex = getSelectedIndex();
		Animation animation = AnimationUtils.loadAnimation(FBStoreActivity.this, playIndex != -1 ? R.anim.panel_in : R.anim.panel_out);
		animation.setAnimationListener(new Animation.AnimationListener() {
			@Override
			public void onAnimationStart(Animation animation) {
				b.panelPack.setVisibility(View.VISIBLE);
				b.panelPack.setAlpha(1);
			}

			@Override
			public void onAnimationEnd(Animation animation) {
				b.panelPack.setVisibility(playIndex != -1 ? View.VISIBLE : View.INVISIBLE);
				b.panelPack.setAlpha(playIndex != -1 ? 1 : 0);
			}

			@Override
			public void onAnimationRepeat(Animation animation) {

			}
		});

		if (playIndex == -1)
			updatePanelMain();
		else
			updatePanelPack(list.get(playIndex));

		int visibility = b.panelPack.getVisibility();
		if ((visibility == View.VISIBLE && playIndex == -1)
				|| (visibility == View.INVISIBLE && playIndex != -1))
			b.panelPack.startAnimation(animation);
	}

	void updatePanelMain() {
		Log.test("panel main");
		b.panelTotal.b.customLogo.setImageResource(R.drawable.custom_logo);
		b.panelTotal.b.version.setText(BuildConfig.VERSION_NAME);
		b.panelTotal.b.storeCount.setText(list.size() + "");
		b.panelTotal.b.downloadedCount.setText(getDownloadedCount() + "");
	}

	void updatePanelPack(StoreItem item) {
		Log.test("panel pack");
		StoreVO storeVO = item.getStoreVO();

		b.panelPack.updateTitle(storeVO.title);
		b.panelPack.updateSubTitle(storeVO.producerName);
		b.panelPack.updateDownloadCount(storeVO.downloadCount);
	}

	// ============================================================================================= Activity

	@Override
	public void onBackPressed() {
		if (getSelectedIndex() != -1)
			togglePlay(null);
		else {
			if (getDownloadingCount() > 0)
				showToast(R.string.canNotQuitWhileDownloading);
			else
				super.onBackPressed();
		}
	}

	@Override
	public void onResume() {
		super.onResume();
		initVar(false);

		list.clear();
		adapter.notifyDataSetChanged();

		togglePlay(null);
		updatePanel();

		firebase_store.attachEventListener(true);
		firebase_storeCount.attachEventListener(true);
	}

	@Override
	public void onPause() {
		super.onPause();

		firebase_store.attachEventListener(false);
		firebase_storeCount.attachEventListener(false);
	}
}

File: app/src/main/java/com/kimjisub/launchpad/activity/FBStoreActivity.kt
package com.kimjisub.launchpad.activity

import android.os.Bundle
import androidx.activity.compose.BackHandler
import androidx.activity.compose.setContent
import androidx.compose.animation.Crossfade
import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.lifecycleScope
import com.google.firebase.database.ChildEventListener
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.kimjisub.launchpad.BuildConfig
import com.kimjisub.launchpad.R
import com.kimjisub.launchpad.adapter.UniPackItem
import com.kimjisub.launchpad.manager.FileManager
import com.kimjisub.launchpad.network.Networks.FirebaseManager
import com.kimjisub.launchpad.network.fb.StoreVO
import com.kimjisub.launchpad.tool.Log
import com.kimjisub.launchpad.tool.UniPackDownloader
import com.kimjisub.launchpad.tool.splitties.browse
import com.kimjisub.launchpad.ui.compose.StorePackPanelContent
import com.kimjisub.launchpad.ui.compose.StoreTotalPanelContent
import com.kimjisub.launchpad.ui.compose.UnipackListItem
import com.kimjisub.launchpad.ui.theme.Gray1
import com.kimjisub.launchpad.ui.theme.Green
import com.kimjisub.launchpad.ui.theme.Orange
import com.kimjisub.launchpad.ui.theme.Red
import com.kimjisub.launchpad.ui.theme.SkyBlue
import com.kimjisub.launchpad.ui.theme.UniPadTheme
import com.kimjisub.launchpad.unipack.UniPack
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import com.google.android.material.snackbar.Snackbar
import java.io.File
import java.text.NumberFormat

class StoreItemState(
	storeVO: StoreVO,
	downloaded: Boolean = false,
) {
	var storeVO by mutableStateOf(storeVO)
	var downloaded by mutableStateOf(downloaded)
	var downloading by mutableStateOf(false)
	var isToggle by mutableStateOf(false)
	var playText by mutableStateOf("")
	var flagColorOverride by mutableStateOf<Color?>(null)
}

class FBStoreActivity : BaseActivity() {
	companion object {
		private const val DOWNLOAD_BASE_URL = "https://us-central1-unipad-e41ab.cloudfunctions.net/downloadUniPackLegacy"
	}

	private val firebaseStore: FirebaseManager by lazy { FirebaseManager("store") }
	private val firebaseStoreCount: FirebaseManager by lazy { FirebaseManager("storeCount") }
	private val storeItems = mutableStateListOf<StoreItemState>()
	private var downloadList: List<UniPackItem> = emptyList()

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		lifecycleScope.launch(Dispatchers.IO) {
			downloadList = ws.getUnipacks()
		}

		setContent {
			UniPadTheme {
				val selectedItem = storeItems.firstOrNull { it.isToggle }
				val hasDownloading = storeItems.any { it.downloading }

				BackHandler {
					if (selectedItem != null) {
						togglePlay(null)
					} else if (hasDownloading) {
						Snackbar.make(findViewById(android.R.id.content), R.string.canNotQuitWhileDownloading, Snackbar.LENGTH_SHORT).show()
					} else {
						finish()
					}
				}

				StoreScreen(
					storeItems = storeItems,
					selectedItem = selectedItem,
					onBackClick = { finish() },
					storeCount = storeItems.size,
					downloadedCount = storeItems.count { it.downloaded },
					onItemClick = { item -> togglePlay(item) },
					onPlayClick = { item ->
						if (!item.downloaded && !item.downloading) {
							startDownload(item)
						}
					},
					onDownloadClick = { item ->
						if (!item.downloaded && !item.downloading) {
							startDownload(item)
						}
					},
					onYoutubeClick = {
						val item = storeItems.firstOrNull { it.isToggle }
						if (item != null) {
							browse("https://www.youtube.com/results?search_query=UniPad+${item.storeVO.title}+${item.storeVO.producerName}")
						}
					},
					onWebsiteClick = {},
				)
			}
		}

		firebaseStore.setEventListener(object : ChildEventListener {
			override fun onChildAdded(dataSnapshot: DataSnapshot, s: String?) {
				try {
					val d: StoreVO = dataSnapshot.getValue(StoreVO::class.java) ?: return
					val isDownloaded = downloadList.any { it.unipack.id == d.code }
					storeItems.add(0, StoreItemState(d, isDownloaded))
				} catch (e: RuntimeException) {
					Log.err("onChildAdded failed", e)
				}
			}

			override fun onChildChanged(dataSnapshot: DataSnapshot, s: String?) {
				try {
					val d: StoreVO = dataSnapshot.getValue(StoreVO::class.java) ?: return
					val code = d.code ?: return
					val item = storeItems.firstOrNull { it.storeVO.code == code } ?: return
					item.storeVO = d
				} catch (e: RuntimeException) {
					Log.err("onChildChanged failed", e)
				}
			}

			override fun onChildRemoved(dataSnapshot: DataSnapshot) {}
			override fun onChildMoved(dataSnapshot: DataSnapshot, s: String?) {}
			override fun onCancelled(databaseError: DatabaseError) {}
		})

		firebaseStoreCount.setEventListener(object : ValueEventListener {
			override fun onDataChange(dataSnapshot: DataSnapshot) {
				val data: Long = dataSnapshot.getValue(Long::class.java) ?: return
				p.prevStoreCount = data
			}

			override fun onCancelled(databaseError: DatabaseError) {}
		})

		firebaseStore.attachEventListener(true)
		firebaseStoreCount.attachEventListener(true)
	}

	private fun togglePlay(target: StoreItemState?) {
		for (item in storeItems) {
			if (target != null && item.storeVO.code == target.storeVO.code)
				item.isToggle = !item.isToggle
			else
				item.isToggle = false
		}
	}

	private fun startDownload(item: StoreItemState) {
		val itemCode = item.storeVO.code ?: return
		item.flagColorOverride = Gray1
		item.playText = "0%"
		item.downloading = true

		UniPackDownloader(
			context = this,
			title = item.storeVO.title ?: "",
			url = "$DOWNLOAD_BASE_URL?code=$itemCode",
			workspace = ws.downloadWorkspace.file,
			folderName = item.storeVO.code ?: "",
			listener = object : UniPackDownloader.Listener {
				override fun onInstallStart() {}

				override fun onGetFileSize(
					fileSize: Long,
					contentLength: Long,
					preKnownFileSize: Long,
				) {
					val downloadedMB: String = FileManager.byteToMB(0)
					val fileSizeMB: String = FileManager.byteToMB(fileSize)
					item.playText = getString(R.string.download_progress_format, 0, downloadedMB, fileSizeMB)
				}

				override fun onDownloadProgress(
					percent: Int,
					downloadedSize: Long,
					fileSize: Long,
				) {
					val downloadedMB: String = FileManager.byteToMB(downloadedSize)
					val fileSizeMB: String = FileManager.byteToMB(fileSize)
					item.playText = getString(R.string.download_progress_format, percent, downloadedMB, fileSizeMB)
				}

				override fun onDownloadProgressPercent(
					percent: Int,
					downloadedSize: Long,
					fileSize: Long,
				) {}

				override fun onImportStart(zip: File) {
					item.playText = getString(R.string.importing)
					item.flagColorOverride = Orange
				}

				override fun onInstallComplete(folder: File, unipack: UniPack) {
					item.playText = getString(R.string.downloaded)
					item.flagColorOverride = Green
					item.downloading = false
					item.downloaded = true
				}

				override fun onException(throwable: Throwable) {
					Log.err("Store download failed", throwable)
					item.playText = getString(R.string.failed)
					item.flagColorOverride = Red
					item.downloading = false
				}
			},
			scope = lifecycleScope,
		)
	}

	override fun onDestroy() {
		firebaseStore.attachEventListener(false)
		firebaseStoreCount.attachEventListener(false)
		super.onDestroy()
	}
}


private fun parseDownloadPercent(playText: String): Float {
	val match = Regex("""^(\d+)%""").find(playText)
	return (match?.groupValues?.get(1)?.toFloatOrNull() ?: 0f) / 100f
}

private const val FLAG_ANIMATION_MS = 500

@Composable
private fun StoreScreen(
	storeItems: List<StoreItemState>,
	selectedItem: StoreItemState?,
	onBackClick: () -> Unit,
	storeCount: Int,
	downloadedCount: Int,
	onItemClick: (StoreItemState) -> Unit,
	onPlayClick: (StoreItemState) -> Unit,
	onDownloadClick: (StoreItemState) -> Unit,
	onYoutubeClick: () -> Unit,
	onWebsiteClick: () -> Unit,
) {
	Row(
		modifier = Modifier
			.fillMaxSize()
			.background(MaterialTheme.colorScheme.background),
	) {
		// Left panel (weight 2)
		Box(
			modifier = Modifier
				.weight(2f)
				.fillMaxHeight()
				.padding(start = 16.dp, top = 16.dp, bottom = 16.dp),
		) {
			Crossfade(
				targetState = selectedItem != null,
				label = "panel",
			) { showPack ->
				if (showPack) {
					val item = storeItems.firstOrNull { it.isToggle }
					if (item != null) {
						StorePackPanelContent(
							title = item.storeVO.title ?: "",
							subtitle = item.storeVO.producerName ?: "",
							downloadCount = NumberFormat.getInstance()
								.format(item.storeVO.downloadCount.toLong()),
							isLED = item.storeVO.isLED,
							isAutoPlay = item.storeVO.isAutoPlay,
							isDownloaded = item.downloaded,
							isDownloading = item.downloading,
							downloadProgress = parseDownloadPercent(item.playText),
							downloadStatusText = item.playText,
							onDownloadClick = { onDownloadClick(item) },
							onYoutubeClick = onYoutubeClick,
							onWebsiteClick = onWebsiteClick,
						)
					}
				} else {
					StoreTotalPanelContent(
						version = BuildConfig.VERSION_NAME,
						storeCount = storeCount,
						downloadedCount = downloadedCount,
					)
				}
			}
		}

		// Right panel (weight 3)
		Column(
			modifier = Modifier
				.weight(3f)
				.fillMaxHeight(),
		) {
			// Store header
			Row(
				modifier = Modifier
					.fillMaxWidth()
					.padding(start = 4.dp, top = 8.dp, end = 16.dp, bottom = 4.dp),
				verticalAlignment = Alignment.CenterVertically,
			) {
				IconButton(onClick = onBackClick) {
					Icon(
						imageVector = Icons.AutoMirrored.Filled.ArrowBack,
						contentDescription = null,
						tint = Color.White,
					)
				}
				Icon(
					painter = painterResource(R.drawable.baseline_shopping_basket_white_24),
					contentDescription = null,
					modifier = Modifier.size(20.dp),
					tint = SkyBlue,
				)
				Spacer(Modifier.width(8.dp))
				Text(
					text = stringResource(R.string.store),
					fontSize = 16.sp,
					fontWeight = FontWeight.Bold,
					color = Color.White,
				)
				Spacer(Modifier.width(8.dp))
				Text(
					text = storeItems.size.toString(),
					fontSize = 12.sp,
					color = Gray1,
				)
			}

			if (storeItems.isEmpty()) {
				Column(
					modifier = Modifier
						.weight(1f)
						.fillMaxWidth(),
					horizontalAlignment = Alignment.CenterHorizontally,
					verticalArrangement = Arrangement.Center,
				) {
					Icon(
						painter = painterResource(R.drawable.ic_error),
						contentDescription = null,
						modifier = Modifier.size(80.dp),
						tint = Gray1,
					)
					Spacer(Modifier.height(8.dp))
					Text(
						text = stringResource(R.string.UnableToAccessServer),
						color = Gray1,
						fontSize = 13.sp,
					)
				}
			} else {
				LazyColumn(
					modifier = Modifier.weight(1f),
					contentPadding = PaddingValues(top = 8.dp, bottom = 6.dp),
					verticalArrangement = Arrangement.spacedBy(8.dp),
				) {
					itemsIndexed(
						storeItems,
						key = { index, item -> item.storeVO.code ?: index },
					) { _, item ->
						StorePackListItem(
							item = item,
							onItemClick = { onItemClick(item) },
							onPlayClick = { onPlayClick(item) },
						)
					}
				}
			}
		}
	}
}

@Composable
private fun StorePackListItem(
	item: StoreItemState,
	onItemClick: () -> Unit,
	onPlayClick: () -> Unit,
) {
	val defaultColor = if (item.downloaded) Green else Red
	val flagWidth by animateDpAsState(
		targetValue = if (item.isToggle) 100.dp else 10.dp,
		animationSpec = tween(FLAG_ANIMATION_MS),
		label = "flagWidth",
	)
	val animatedFlagColor by animateColorAsState(
		targetValue = item.flagColorOverride ?: defaultColor,
		animationSpec = tween(FLAG_ANIMATION_MS),
		label = "flagColor",
	)

	UnipackListItem(
		title = item.storeVO.title ?: "",
		subtitle = item.storeVO.producerName ?: "",
		hasLed = item.storeVO.isLED,
		hasAutoPlay = item.storeVO.isAutoPlay,
		indicatorFontSize = 12.sp,
		flagColor = animatedFlagColor,
		flagWidth = flagWidth,
		contentStartPadding = flagWidth,
		flagClickable = item.isToggle,
		onFlagClick = onPlayClick,
		flagContent = {
			if (item.isToggle) {
				Text(
					text = item.playText.ifEmpty {
						stringResource(if (item.downloaded) R.string.downloaded else R.string.download)
					},
					fontSize = 12.sp,
					textAlign = TextAlign.Center,
					maxLines = 1,
				)
			}
		},
		onClick = onItemClick,
	)
}
